File: javascript/binary-search.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function input(question) {
  return new Promise((resolve) => {
    rl.question(question, resolve);
  });
}

function binarySearch(numbers, number) {
  let start = 0;
  let end = numbers.length;
  let mid = Math.floor((start + end) / 2);

  while (start <= end) {
    if (numbers[mid] === number) {
      return mid;
    } else if (numbers[mid] > number) {
      end = mid - 1;
    } else {
      start = mid + 1;
    }

    mid = Math.floor((start + end) / 2);
  }

  throw new Error('Number not found!');
}

async function main() {
  const numbers = [];
  let number = 0;
  const size = parseInt(await input("Enter array size: "));

  console.log("Enter array values:");

  for (let i = 0; i < size; i++) {
    number = parseInt(await input(""));
    numbers.push(number);
  }

  number = parseInt(await input("Enter number to search: "));

  rl.close();

  console.log(`Unsorted: ${numbers}`);

  numbers.sort((a, b) => a - b);

  console.log(`Sorted: ${numbers}`);

  try {
    const index = binarySearch(numbers, number);

    console.log(`Index of ${number}: ${index}`);
  } catch (error) {
    console.log(error.message);
  }
}

main();

File: go/BinarySearch.go
package main

import (
	"slices"
	"errors"
	"fmt"
)

func binarySearch(numbers []int, number int) (int, error) {
	start := 0
	end := len(numbers) - 1
	mid := (end + start) / 2

	for start <= end {
		if numbers[mid] == number {
			return mid, nil
		} else if numbers[mid] > number {
			end = mid - 1
		} else {
			start = mid + 1
		}

		mid = (end + start) / 2
	}

	return -1, errors.New("Number not found!")
}

func main() {
	// numbers := []int{ 4, 12, 100, 50, 32 }
	// number := 50
	var size int
	var num int
	var number int
	numbers := []int{}

	fmt.Println("Enter the size of the array:")
	fmt.Scanln(&size)

	fmt.Println("Enter array values:")

	for i := 0; i < size; i++ {
		fmt.Scanln(&num)
		numbers = append(numbers, num)
	}

	fmt.Println("Enter number to search:")
	fmt.Scanf("%d", &number)

	fmt.Printf("Unsorted: %v\n", numbers)

	slices.Sort(numbers)

	fmt.Printf("Sorted: %v\n", numbers)

	index, err := binarySearch(numbers, number)

	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Printf("Index of %v is: %v", number, index)
	}
}